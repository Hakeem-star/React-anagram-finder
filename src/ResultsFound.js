import React from "react";

export default class ResultsFound extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      currentResultsPage: 0,
      maxPages: 1
    };
  }

  //Seperate the results into chunks, so we can have different pages for the results
  chunk(arr, chunckSize) {
    // debugger;
    let newChunk = [];
    let j = 0;
    for (let i = 0; i < arr.length; i += chunckSize) {
      !i ? (j = 0) : (j += chunckSize);
      newChunk.push(arr.slice(j, j + chunckSize));
    }

    return newChunk;
  }

  navigateResults(e) {
    if (e.target.textContent === " » ") {
      if (this.state.currentResultsPage < this.state.maxPages - 1) {
        this.setState({
          currentResultsPage: this.state.currentResultsPage + 1
        });
      }
    } else if (e.target.textContent === " « ") {
      if (this.state.currentResultsPage > 0) {
        this.setState({
          currentResultsPage: this.state.currentResultsPage - 1
        });
      }
    }
    console.log(this.state.currentResultsPage, this.state.maxPages - 1);
    console.dir(e.target.textContent);
  }

  // componentDidMount() {
  //   //console.log( this.chunk(Object.keys(this.props.results),10));
  //   this.setState({ maxPages: this.chunk(Object.keys(this.props.results),10).length });
  // }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.results !== this.props.results) {
      this.setState({
        maxPages: this.chunk(Object.keys(this.props.results), 10).length
      });
    }
  }

  render() {
    let rows = Array(Object.keys(this.props.results).length)
      .fill("")
      .map((value, ind) => {
        const anagramName = Object.keys(this.props.results)[ind]; //Thanksmo
        const val = this.props.results[anagramName][0]; //return the array associated with the anagram
        /*0: {Tom Hanks: "100%"} */
        const celebName = Object.keys(val)[0]; //Create an array based off the keys in the returned object.
        //"Tom Hanks"
        const celebNameURL = (
          <a target="blank" href={"https://en.wikipedia.org/wiki/" + celebName}>
            {celebName}
          </a>
        ); //Create a wiki link based off the celebname.
        const innerCurrent = val[celebName];
        return (
          <tr key={anagramName + celebName}>
            <td>{anagramName}</td>
            <td>{celebNameURL}</td>
            <td>{innerCurrent}</td>
          </tr>
        );
      });

    let resultsCounter =
      this.state.currentResultsPage > 0
        ? this.chunk(rows, 10)[this.state.currentResultsPage].length +
          this.state.currentResultsPage * 10
        : this.chunk(rows, 10)[this.state.currentResultsPage].length;

    return (
      <div className="container">
        <table>
          <thead>
            <tr>
              <th>Anagram</th>
              <th>Name</th>
              <th>Match Rate</th>
            </tr>
          </thead>
          <tbody>{this.chunk(rows, 10)[this.state.currentResultsPage]}</tbody>
        </table>
        <ResultsPageNumber
          prevResults={e => this.navigateResults(e)}
          current={resultsCounter}
          nextResults={e => this.navigateResults(e)}
          maximum={rows.length}
        />
      </div>
    );
  }
}

export function ResultsPageNumber(props) {
  return (
    <div className="ResultsPageNumber">
      <p className="previous" onClick={props.prevResults}>
        {" "}
        «{" "}
      </p>
      <p>
        Showing {props.current} of {props.maximum}
      </p>
      <p className="next" onClick={props.nextResults}>
        {" "}
        »{" "}
      </p>
    </div>
  );
}
