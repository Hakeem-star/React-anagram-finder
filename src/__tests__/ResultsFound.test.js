import React from "react";
import ResultsFound from "../ResultsFound";
import toJson from "enzyme-to-json";
import { mount } from "enzyme";

let resultsProps = JSON.parse(
  JSON.stringify({
    THANKSMO: [{ "Tom Hanks": "100%" }],
    TRUMCDONALD: [{ "Donald Trump": "100%" }],
    THANKSMO2: [{ "Tom Hanks": "100%" }],
    TRUMCDONALD2: [{ "Donald Trump": "100%" }],
    THANKSMO3: [{ "Tom Hanks": "100%" }],
    TRUMCDONALD3: [{ "Donald Trump": "100%" }],
    THANKSMO4: [{ "Tom Hanks": "100%" }],
    TRUMCDONALD4: [{ "Donald Trump": "100%" }],
    THANKSMO5: [{ "Tom Hanks": "100%" }],
    TRUMCDONALD5: [{ "Donald Trump": "100%" }],
    THANKSMO6: [{ "Tom Hanks": "100%" }],
    TRUMCDONALD6: [{ "Donald Trump": "100%" }]
  })
);

describe("<ResultsFound />", () => {
  let wrapper;

  it("matches the snapshot", () => {
    wrapper = mount(<ResultsFound results={resultsProps} />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });

  it("lets <ResultsPageNumber/> navigate pages properly", () => {
    //const onButtonClick = sinon.spy();
    //const wrapper = shallow(<MyComponent />);
    wrapper = mount(<ResultsFound results={resultsProps} />);
    console.log(wrapper.state("currentResultsPage"));
    //wrapper.setState({ name: "bar" });
    wrapper.find("next").simulate("click");
    expect(wrapper.state("currentResultsPage")).to.equal(1);
  });
});
