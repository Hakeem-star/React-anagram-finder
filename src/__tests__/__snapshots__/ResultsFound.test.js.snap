// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ResultsFound /> matches the snapshot 1`] = `
<ResultsFound
  results={
    Object {
      "THANKSMO": Array [
        Object {
          "Tom Hanks": "100%",
        },
      ],
      "THANKSMO2": Array [
        Object {
          "Tom Hanks": "100%",
        },
      ],
      "THANKSMO3": Array [
        Object {
          "Tom Hanks": "100%",
        },
      ],
      "THANKSMO4": Array [
        Object {
          "Tom Hanks": "100%",
        },
      ],
      "THANKSMO5": Array [
        Object {
          "Tom Hanks": "100%",
        },
      ],
      "THANKSMO6": Array [
        Object {
          "Tom Hanks": "100%",
        },
      ],
      "TRUMCDONALD": Array [
        Object {
          "Donald Trump": "100%",
        },
      ],
      "TRUMCDONALD2": Array [
        Object {
          "Donald Trump": "100%",
        },
      ],
      "TRUMCDONALD3": Array [
        Object {
          "Donald Trump": "100%",
        },
      ],
      "TRUMCDONALD4": Array [
        Object {
          "Donald Trump": "100%",
        },
      ],
      "TRUMCDONALD5": Array [
        Object {
          "Donald Trump": "100%",
        },
      ],
      "TRUMCDONALD6": Array [
        Object {
          "Donald Trump": "100%",
        },
      ],
    }
  }
>
  <div
    className="container"
  >
    <table>
      <thead>
        <tr>
          <th>
            Anagram
          </th>
          <th>
            Name
          </th>
          <th>
            Match Rate
          </th>
        </tr>
      </thead>
      <tbody>
        <tr
          key="THANKSMOTom Hanks"
        >
          <td>
            THANKSMO
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Tom Hanks"
              target="blank"
            >
              Tom Hanks
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="TRUMCDONALDDonald Trump"
        >
          <td>
            TRUMCDONALD
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Donald Trump"
              target="blank"
            >
              Donald Trump
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="THANKSMO2Tom Hanks"
        >
          <td>
            THANKSMO2
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Tom Hanks"
              target="blank"
            >
              Tom Hanks
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="TRUMCDONALD2Donald Trump"
        >
          <td>
            TRUMCDONALD2
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Donald Trump"
              target="blank"
            >
              Donald Trump
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="THANKSMO3Tom Hanks"
        >
          <td>
            THANKSMO3
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Tom Hanks"
              target="blank"
            >
              Tom Hanks
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="TRUMCDONALD3Donald Trump"
        >
          <td>
            TRUMCDONALD3
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Donald Trump"
              target="blank"
            >
              Donald Trump
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="THANKSMO4Tom Hanks"
        >
          <td>
            THANKSMO4
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Tom Hanks"
              target="blank"
            >
              Tom Hanks
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="TRUMCDONALD4Donald Trump"
        >
          <td>
            TRUMCDONALD4
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Donald Trump"
              target="blank"
            >
              Donald Trump
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="THANKSMO5Tom Hanks"
        >
          <td>
            THANKSMO5
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Tom Hanks"
              target="blank"
            >
              Tom Hanks
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
        <tr
          key="TRUMCDONALD5Donald Trump"
        >
          <td>
            TRUMCDONALD5
          </td>
          <td>
            <a
              href="https://en.wikipedia.org/wiki/Donald Trump"
              target="blank"
            >
              Donald Trump
            </a>
          </td>
          <td>
            100%
          </td>
        </tr>
      </tbody>
    </table>
    <ResultsPageNumber
      current={10}
      maximum={12}
      nextResults={[Function]}
      prevResults={[Function]}
    >
      <div
        className="ResultsPageNumber"
      >
        <p
          className="previous"
          onClick={[Function]}
        >
           
          «
           
        </p>
        <p>
          Showing 
          10
           of 
          12
        </p>
        <p
          className="next"
          onClick={[Function]}
        >
           
          »
           
        </p>
      </div>
    </ResultsPageNumber>
  </div>
</ResultsFound>
`;
